designs/auto_scaling/VNF Autoscaling.md

```diff
@@ -0,0 +1,91 @@
+
+We have 2 Basic options
+1. Create a VNF deployment tied to the number hosts within a Availability Zone/region/Host Aggregate etc
+Auto scaling is “built-in” when new hosts get added/deleted VNF also get added/deleted (there is still work that needs to be done to enable this in OSC)
+
+2. Static definition of number of VNF’s per host.
+Here we select the hosts where we want VNF’s to be deployed AND we select the number of instances to be deployed per host. <br/>
+Auto-scaling here means we increase the number of instances while keeping the hosts constant. The intent is the security admin is providing a base line from which you would scale up and scale down.
+
+## Constraints and Assumptions
+The Security manager is monitoring the appliance instances and can determine if a particular instance is overloaded.
+
+## Design Changes
+
+### Deployment
+The admin provides the lower threshold implicitly when creating the deployment specification. In addition, we take an option to specify the upper threshold for the number of instance per host. When a scaling up request comes in, we will deploy extra VNF's upto the upper threshold defined by the administrator.
+
+For example, the deployment specifiation can declare that for 2 hosts H1 and H2 we will deploy 1 VNF instance and the upper theshold is set to 3 VNF instances.
+
+When a scaling request comes up, we are expecting to get the [exact instance which is overloaded as part of the request.](#Constraints-and-Assumptions) Assume H1/VNF1 is this instace <br/>
+OSC will deploy H1/VNF2 the first time and an additional H1/VNF3 if another scaling up request comes in. OSC should NOT deploy any additonal VNF's in H2 since that VNF did not initiate a scale request.
+If another scaling request comes in, OSC has exhausted the scale up policy defined and throws an error.<br/>
+The scale down happens the same way, the extra VNF's are undeployed until we are left with the baseline VNF's only. 
+Any further scale down request will result in a no-op.
+
+### Load balance/HA redundancy
+
+Once the VNF's are deployed/undeployed, it is upto the SDN controller to take advantage of the extra VNF's.
+
+When a new VNF's is deployed in response to a scale event, OSC registers this new VNF with the SDN controller(as well as the manager) under a "Port Group" where the existing VNF's are already registered. The SDN controller determines how the traffic is load balanced between all the instances. There are a variety of ways for the networking layer to implement this starting with a basic round robin approach to using a consistent hashing mechanism using the port id's/mac/IP.
+
+This functionality is also relavent to providing High Availability/Redundancy to the sercurity functions. If one of the VNF instances goes down another one in the same port group will pick up the traffic. If this causes extra load, the scale up request can provide additional VNF instances as necessary.
+
+![](./images/Port_Group.png)
```

Comment:
we are looking for feedback. My understanding is while nuage does not load balancing at this time, it is definitely something they are looking at to support.